### PR TITLE
Add connection loss error hook

### LINKED ISSUE #37: Expand on connectivity lifecycle
Following https://github.com/featherbear/presonus-studiolive-api/pull/35#issuecomment-1312609587, the following features should be added to the roadmap

- [x] Auto-reconnect - Allow the client to reconnect on disconnect.
- [x] Lock down on event mutations
- [x] Refactor vendor types

Also, on connection loss via disconnection (i.e. wifi disruption), the error event is only raised after about a minute.
We should investigate on finding a health check, as the KA (keepalive) packet does not pong.

### PR DESCRIPTION
**Pull request by Akkjon**
Hi,
this is my first attempt on creating an error hook for connection losses or other connection errors like timeouts.

Usage would be:
```
const client = new Client('192.168.178.57');

client.on(ConnectionState.Closed, () => {
  console.log("closed")
})

client.on(ConnectionState.Error, (e) => {
  console.log("error", e)
})

client.connect().then(() => {
    // do something
  }).catch(()=> {
    // catch initial connection error, produced by promise rejection
  })
}
```

If the catch statement after the connect method, even if on(error) is hooked, the application crashes, because  of a promise rejection error.If the error occurs before the first connection is possible, the catch code is executed as well.
But regardless of when the error occurs, the error function is always called, I tried it on a StudioLive 32R mixer, there it works perfectly.
The close method is called every time the connection to the mixer is closed, so if successfully closed, or closed by a network error.

If you have any comments on that, please let me know. I am thinking if there is anything to improve with the solution I implemented, thats why the pull request is still a draft :)

### CODE REVIEW

**Comment by Akkjon** [comment]
I don't find any way to make the thing "pretier".
Maybe removing the necessity to include the catch, so resolve the promise, even if the connection is not successful, but that seems more like a "hack" to me...

**Comment by Akkjon** [comment]
If the process gets paused, the socket stays connected but the interval for the keepalive packet stops.
That means if for example windows goes to sleep, the session is closed by the mixer, because no keep alive packets are sent but no connection loss is handled because the process keeps sending keep alive packets to the mixer nevertheless.

**Comment by featherbear** [comment]
Thanks for that, having a look at it now

Do you have a use case for wanting to receive the disconnect event (feedback?), else it might be better to automatically reconnect from within the client instead of requiring the user to reconnect

**Comment by Akkjon** [comment]
Ok, valid question, thanks.

TLDR: To set states on a connected midi controller when the mixer is turned off, reconnect would not make sence when the mixer is turned off

Long version:
I have connected a Launchpad (MIDI device) to the computer, this device has buttons that let me control mute states and levels of selected channels. This software runs all the time, the computer starts up. It starts a discovery server and when connected, it loads the buttons colors up. When the mixer gets disconnected, it will (in the future) restart the discovery server and turn off the launchpad buttons. So if the mixer is turned off, an automatic reconnect would fail and would not make sence for my personal use case. 

i hope, that makes sence to you, if not: let me know

File: src/lib/Client.ts [hunk 1]
```diff
@@ -163,12 +167,14 @@ export class Client extends EventEmitter {
...
-        this.connectPromise = null
+      const rejectHandler = (connectionState: ConnectionState) => (err: Error) => {
+        this.connectPromise = null;
+        this.emit(connectionState, err);
         return reject(err)
       }
 
-      this.conn.once('error', rejectHandler)
+      this.conn.once('error', rejectHandler(ConnectionState.Error))
+      this.conn.once('close', rejectHandler(ConnectionState.Closed))
```
Comment (featherbear) [inline src/lib/Client.ts:181]: Do both events always occur? Otherwise we may have a stacked callback leak here

Comment (Akkjon) [reply src/lib/Client.ts:181]: close is always called, error is only called, when the connection is closed because of a connection error.

Comment (featherbear) [reply src/lib/Client.ts:181]: Okay, all the events will need be unregistered if an event is thrown - otherwise after a loop of connections and closes, if an error occurs it will trigger the error event multiple times

File: src/lib/Client.ts [hunk 2]
```diff
@@ -233,11 +239,20 @@ export class Client extends EventEmitter {
 
         // #region Keep alive
         // Send a KeepAlive packet every second
+        let lastKeepAlive = new Date().getTime();
```
Comment (featherbear) [inline src/lib/Client.ts:246]: I don't think this keep-alive check is functional, there's no condition that won't cause it to be updated every interval

Comment (featherbear) [reply src/lib/Client.ts:246]: image

Comment (Akkjon) [reply src/lib/Client.ts:246]: I see your point. But at some point in testing, I got the error that when my computer (Windows) got to sleep, the connection was still not "disconnected", but the mixer did close the connection because no keep-alive got send in that time.
This problem is handled here, if the program is haltet by the operating system and is resumed later on, the time stamps differ. I should probably add a comment for this in the code...

File: src/lib/Client.ts (hunk 1 above)
Comment (Akkjon) [reply src/lib/Client.ts:181]: Implemented in commit 48e32fc9850885f2f6cb14bf796c0db4478757a3

File: src/lib/Client.ts (hunk 2 above)
Comment (featherbear) [reply src/lib/Client.ts:246]: Right, got it!

**Comment by featherbear** [comment]
I'm not too sure about the design choice to create `constants/connection.ts` because all the files in that same directory are vendor constants... maybe I'll need to create a `vendor` folder.. 

I noticed that if the error event is not subscribed to, then the reject handler will crash the program (Node v15 iirc) and the close event will not fire. I think the error event should probably be optionally subscribed to, as the close event should have a higher precedence.

I'm also abit tentative with having a `removeAllListeners` call as it would remove listeners that aren't from this library.

Otherwise looks good, thanks for your help!
I'll merge this into a dev branch for the time being and clean things up
